src/librustc/hir/mod.rs

```diff
@@ -1450,8 +1456,7 @@ pub enum Ty_ {
     /// where `Bound` is a trait or a lifetime.
     TyTraitObject(HirVec<PolyTraitRef>, Lifetime),
     /// An `impl Bound1 + Bound2 + Bound3` type
-    /// where `Bound` is a trait or a lifetime.
-    TyImplTrait(TyParamBounds),
+    TyImplTrait(ExistTy, HirVec<Lifetime>),
```

Comment:
Why are the lifetimes separate? Shouldn't they go in the `TyParamBounds`?

Reply:
Nope. These are the lifetimes that are in the *reference* to the abstract type, not its definition. That is, consider the desugaring of `fn foo<'a>() -> impl Trait<'a>`. You get:

```rust
abstract type Foo<'b>: Trait<'b>;
//                ^^ this `'b` is what appears in the `ExistTy`
//                           ^^ and this is the `TyParamBounds`, note that
//                              the lifetime here will reference the lifetime 
//                              `'b` defined in the `abstract type`
fn foo<'a>() -> Foo<'a> { .. }
//                  ^^ this `'a` is the one that appears in this `HirVec<Lifetime>
```

(Clearly, we need some more comments here.)

Reply:
These lifetimes are resolved outside of the generics, so they pick up whatever is in the environment-- see [here](https://github.com/rust-lang/rust/pull/45701/files/a76cc757c1ef519f94a15fd27e6112ade4308ebe#diff-9694c2dab1b4140752e569e4b1ba58c4R420). Later, those lifetimes are "applied" to the generics [here](https://github.com/rust-lang/rust/pull/45701/files/a76cc757c1ef519f94a15fd27e6112ade4308ebe#diff-1b4d50b82a87f1031cd5f99b64123de7R1036).

You can think about a `TyImplTrait` as an `abstract type Foo<'a, 'b>: MyTrait + 'a + 'b;` paired with the lifetime parameters applied to it, `'a` and `'b`. You can also think about this list as containing all of the type parameters, but they don't have to be separately applied since they're just picked up from the parent generics.

Reply:
r=me with a comment for this added.